### PR TITLE
Enable ADL for Monsters

### PR DESCRIPTION
**Pull request by obligaron**
Contributes to #838

<details><summary>Example Video</summary>

### Master

https://user-images.githubusercontent.com/25415264/122987556-04e14680-d3a1-11eb-99ac-87ad25775cc7.mp4

### PR

https://user-images.githubusercontent.com/25415264/122987578-0b6fbe00-d3a1-11eb-81ff-359493dbb465.mp4

</details>

Notes:

- This PR introduces ADL to Monsters
- Currently it's used for
   - Attacks
   - Range-Attacks
   - Spell-Attacks
- The other animations are not included to make the size of the pr manageable.
- This pr lays the groundwork for future followup PRs. Similar to what was done for players:
  - Walking
  - Decouple positioning while walking from game logic
  - Blocking
  - MA_SPECIAL (for example gargoyle heal)
  - Death

### CODE REVIEW

File: Source/engine/animationinfo.cpp [hunk 1]
```diff
@@ -170,18 +171,35 @@ void AnimationInfo::ChangeAnimationData(CelSprite *pCelSprite, int numberOfFrame
 	DelayLen = delayLen;
 }
 
-void AnimationInfo::ProcessAnimation()
+void AnimationInfo::ProcessAnimation(bool reverseAnimation /*= false*/, bool dontProgressAnimation /*= false*/)
 {
 	DelayCounter++;
+	if (dontProgressAnimation)
+		return;
```
Comment (AJenbo) [inline Source/engine/animationinfo.cpp:178]: It's a bit hard to see if the logic behind this is proper with out any usage, maybe wait with adding `dontProgressAnimation` until it has a use case.

File: Source/monster.cpp [hunk 2]
```diff
@@ -5433,7 +5416,13 @@ void decode_enemy(int m, int enemy)
 void MonsterStruct::CheckStandAnimationIsLoaded(int mdir)
 {
 	if (_mmode == MM_STAND || _mmode == MM_TALK)
-		_mAnimData = &*MType->Anims[MA_STAND].CelSpritesForDirections[mdir];
+		AnimInfo.pCelSprite = &*MType->Anims[MA_STAND].CelSpritesForDirections[mdir];
+}
+
+void MonsterStruct::SetStoneMode()
```
Comment (AJenbo) [inline Source/monster.cpp:5422]:
```suggestion
void MonsterStruct::Petrify()
```

File: Source/monster.cpp [hunk 3]
```diff
@@ -5433,7 +5416,13 @@ void decode_enemy(int m, int enemy)
...
 {
 	if (_mmode == MM_STAND || _mmode == MM_TALK)
-		_mAnimData = &*MType->Anims[MA_STAND].CelSpritesForDirections[mdir];
+		AnimInfo.pCelSprite = &*MType->Anims[MA_STAND].CelSpritesForDirections[mdir];
+}
+
+void MonsterStruct::SetStoneMode()
+{
+	_mmode = MM_STONE;
+	AnimInfo.IsFrozen = true;
```
Comment (AJenbo) [inline Source/monster.cpp:5425]:
```suggestion
	AnimInfo.IsPetrified = true;
```

File: Source/missiles.cpp [hunk 4]
```diff
@@ -4422,9 +4422,10 @@ void MI_Stone(int i)
 
 	if (missile[i]._mirange == 0) {
 		missile[i]._miDelFlag = true;
-		if (monster[m]._mhitpoints > 0)
+		if (monster[m]._mhitpoints > 0) {
 			monster[m]._mmode = (MON_MODE)missile[i]._miVar1;
-		else
+			monster[m].AnimInfo.IsFrozen = false;
+		} else
 			AddDead(monster[m].position.tile, stonendx, monster[m]._mdir);
```
Comment (AJenbo) [inline Source/missiles.cpp:4429]:
```suggestion
		} else {
			AddDead(monster[m].position.tile, stonendx, monster[m]._mdir);
		}
```

File: Source/monster.cpp [hunk 5]
```diff
@@ -2298,15 +2295,15 @@ bool M_DoRSpAttack(int i)
 		PlayEffect(i, 3);
 	}
 
-	if (monster[i]._mAi == AI_MEGA && monster[i]._mAnimFrame == 3) {
+	if (monster[i]._mAi == AI_MEGA && monster[i].AnimInfo.CurrentFrame == monster[i].MData->mAFNum2) {
```
Comment (AJenbo) [inline Source/monster.cpp:2298]: Nice :)

**Review by AJenbo — CHANGES_REQUESTED** [review]
Looks good, just a bit of naming and code style.

**Comment by sumyunggai** [comment]
While my observations are those of a total amateur watching professionals work, @AJenbo you're doing a good job of keeping the code readable with your suggested name changes.

File: Source/monster.cpp (hunk 2 above)
Comment (obligaron) [reply Source/monster.cpp]: fixed 🙂

File: Source/monster.cpp (hunk 3 above)
Comment (obligaron) [reply Source/monster.cpp]: fixed 🙂

File: Source/missiles.cpp (hunk 4 above)
Comment (obligaron) [reply Source/missiles.cpp]: fixed 🙂

File: Source/monster.cpp (hunk 5 above)
Comment (obligaron) [reply Source/monster.cpp:2298]: thanks 🙂

**Comment by AJenbo** [comment]
> While my observations are those of a total amateur watching professionals work, @AJenbo you're doing a good job of keeping the code readable with your suggested name changes.

Thanks :)

**Review by AJenbo — APPROVED** [review]

**Pull request merged**